### DISCUSSION TITLE
Attention checks after n trials within randomized timeline

### DISCUSSION THREAD

**Discussion by gaiamolinaro** [jspsych/jsPsych#1076] [Q&A] [answered]
Hi! I have a timeline where I iterate, in random order, through different trials by use of jsPsych.timelineVariable. Attention checks are among a  list of objects called "trials" (with all normal trials). I use timeline_variables: trials, and randomize_order: true. The problem is, I would like these attention check trials to be shown only after N trials (let's say 10 of them), or to appear after a specific number of trials, e.g. trial 10 and trial 22. 
Here's a basic overview of the core of my task's code:
```
var trials = [
  {id: "trial_1",
  stimulus: //directory for image_1.png
},
  {id: "trial_2",
  stimulus: //directory for image_2.png
}, //and so on, and then
{id: "attcheck_1",
  stimulus: //directory for attcheck_1.png
},
{id: "attcheck_2",
  stimulus: //directory for attcheck_2.png
},
 var node = { 
         timeline: [
        {
      type: 'html-keyboard-response',
      stimulus: function() {
       // returning "<img src='" + jsPsych.timelineVariable('stimulus',true) "'>"
      } 
      choices: [choiceX, choiceY]
       }, //other parts of each trial added similarly
         ],
         timeline_variables: trials,
         randomize_order: true,
         } 
     timeline.push(node);
```
Thank you so much in advance! Let me know if you need any further detail to be specified.

**Comment by becky-gilbert** [comment, marked as answer]
Hi @gaiamolinaro, one way to do this is to use the [custom sampling option](https://www.jspsych.org/overview/timeline/#custom-sampling-function) for timeline variables, instead of `randomize_order: true`. This will allow you to take the indices of your original `trials` array and output a new array that specifies your trials in a pseudorandomized order. Let's say that your trial array looks like this, with 4 'real' trials and 2 attention checks:
```javascript
var trials = [
  {id: "trial_1"},
  {id: "trial_2"}, 
  {id: "trial_3"}, 
  {id: "trial_4"}, 
  {id: "attcheck_1"},
  {id: "attcheck_2"}
];
```
Then you could add a custom sampling function to your `node` object like this:
```javascript
sample: {
  type: 'custom',
  fn: function(t) {
    var new_order = [];
    var n_attention_checks = 2;
    var attention_check_spacing = 3; // i.e. attention check is every 3rd trial
    var real_trial_indices = t.slice(0,t.length-n_attention_checks); // [0,1,2,3]
    var attention_check_indices = t.slice(t.length-n_attention_checks); // [4,5]
    // shuffle the order of the real trials
    var real_trial_indices_random = jsPsych.randomization.shuffle(real_trial_indices);
    // loop through the total number of trials and combine the shuffled 'real trial' indices with the attention check indices
    for (var i=0; i<t.length; i++) {
      if (((i+1) % attention_check_spacing) == 0) {
        // if this trial should be an attention check, then select the next element in the attention_check_indices array
        new_order.push(attention_check_indices.shift());
      } else {
        // if this trial should be a real trial, then select the next element from the real_trial_indices_random array
        new_order.push(real_trial_indices_random.shift());
      }
    }
    console.log('new order: ', new_order);
    return new_order;
  }
}
```
The `new_order` array that is returned will specify the order that your trials should run in, where each number represents the original trial index. In this example, the 3rd element (index = 2) and 6th element (index = 5) in `new_order` should always be the numbers '4' and '5', representing the indices of the last two elements of your original `trials` array, which are your two attention checks. The other elements in the `new_order` array should be the numbers 0-3 (i.e. the indices of your 'real' trials) in random order.

Here's what I got in the console when I last ran it:
new order:  Array(6)
0: 0
1: 3
2: 4
3: 2
4: 1
5: 5

**Reply by gaiamolinaro** [reply]
That worked perfectly! Thank you so much for writing your answer so clearly - it helped me to not only solve my problem but also learn something new.
